### PR TITLE
Appnexus/anxr 372 export default stateful

### PR DESCRIPTION
**Pull request by sodiumjoe**
changed PR target of https://github.com/appnexus/lucid/pull/206 to `release/1.0.0`

## PR Checklist

- Manually tested across supported browsers
  - ~~[ ] Chrome~~
  - ~~[ ] Firefox~~
  - ~~[ ] Safari~~
  - ~~[ ] IE11 (Win 7)~~
  - ~~[ ] Edge (Win 10)~~
- ~~[ ] Unit tests written (`common` at minimum)~~
- [x] PR has one of the `semver-` labels
- [ ] Two core team engineer approvals
- ~~[ ] One core team UX approval~~

- [x] https://github.com/appnexus/lucid/pull/206#discussion_r61453813
- [x] https://github.com/appnexus/lucid/pull/206#issuecomment-215273896
- [x] run buildStateful deprecation warning only once
- [ ] finalize "dumb" component suffix

### CODE REVIEW

File: src/components/SingleSelect/SingleSelect.jsx [hunk 1]
```diff
@@ -1,7 +1,7 @@
 import React from 'react';
 import _ from 'lodash';
 import { lucidClassNames } from '../../util/style-helpers';
-import { createLucidComponentDefinition } from '../../util/component-definition';
+import { createClass } from '../../util/component-definition';
 import * as reducers from './SingleSelect.reducers';
 import DropMenu from '../DropMenu/DropMenu';
 import CaretIcon from '../Icon/CaretIcon/CaretIcon';
```
Comment (ogupte) [inline src/components/SingleSelect/SingleSelect.jsx]: are we going to import DropMenu and CarentIcon from `../../../index`?

File: src/components/Table/Table.jsx [hunk 2]
```diff
@@ -1,27 +1,40 @@
 import _ from 'lodash';
 import React from 'react';
 import { lucidClassNames } from '../../util/style-helpers';
-import { createLucidComponentDefinition } from '../../util/component-definition';
+import { createClass } from '../../util/component-definition';
 import { findElementsByType } from '../../util/child-component';
 import CaretIcon from '../Icon/CaretIcon/CaretIcon';
 import DragCaptureZone from '../DragCaptureZone/DragCaptureZone';
```
Comment (ogupte) [inline src/components/Table/Table.jsx]: import CaretIcon, DragCaptureZone from `../../../index` ?

**Comment by ogupte** [comment]
i feel like this change makes the code more confusing overall, some components are being imported by other from `index` some are directly imported, `there are dumb versions of some components, but not others, are we importing the dump or not-dumb versions of components from `index` what if there was no _dumb_ version index which one are we importing then?

I feel that this could be more consistent and simple if we were to have `buildStatefulComponent` devolve in to the identity function if there are no reducers/selectors, and then _always_ hang a `.Dumb` static property off the returned component. then run it thru _all_ exported components.

File: src/util/util.js [hunk 3]
```diff
@@ -0,0 +1,74 @@
+import _ from 'lodash';
```
Comment (ogupte) [inline src/util/util.js]: this file should be named `util/environment.js` or something rather than just the generic `util.js`

Comment (ogupte) [inline src/util/util.js]: also this file is missing tests

File: src/util/state-management.js [hunk 4]
```diff
@@ -82,7 +86,18 @@ function overwriteArrays (objValue, srcValue) {
 	}
 }
 
-export function buildStatefulComponent(baseComponent, opts) {
+export function buildHybridComponent(baseComponent, opts) {
```
Comment (ogupte) [inline src/util/state-management.js]: there's no tests for this function.

File: src/components/SingleSelect/SingleSelect.jsx (hunk 1 above)
Comment (sodiumjoe) [inline src/components/SingleSelect/SingleSelect.jsx]: no, that's the public hybrid version. Internally, we compose using dumb components. Right?

File: src/util/state-management.js [hunk 5]
```diff
@@ -1,5 +1,9 @@
 import React from 'react';
 import _ from 'lodash';
+import { logger } from './util';
+
+let hasLoggedBuildStateComponentWarning = false;
+const alreadyWarnedHybrid = {};
```
Comment (ogupte) [inline src/util/state-management.js]: this log-once type logic should be part of `logger`!

**Comment by sodiumjoe** [comment]
I agree that it's confusing that there are dumb versions of some components and not others. There's not really a good solution to this, though, imo. Even if we run everything through buildHybrid, that means some components have a `hybrid` version but no reducers, and also a dumb version that, to a consumer, is identical. I think it's confusing either way we fall on that one.

on `import`ing from `../../index`, we only do that in the examples, because the examples use the library like a consumer would. Components should only ever use direct references to components (i.e. the dumb versions).

File: src/util/util.js (hunk 3 above)
Comment (sodiumjoe) [inline src/util/util.js]: fixed

File: src/util/state-management.js (hunk 4 above)
Comment (sodiumjoe) [inline src/util/state-management.js]: fixed

File: src/util/state-management.js (hunk 5 above)
Comment (sodiumjoe) [inline src/util/state-management.js]: fixed

**Pull request merged**